### PR TITLE
fix(hnsw): bump hnswlib to v0.6.2

### PR DESCRIPTION
**Pull request by OrangeSodahub**
Update the hnswlib (located in `include`) from https://github.com/nmslib/hnswlib.

### CODE REVIEW

File: include/hnswlib/bruteforce.h [hunk 1]
```diff
@@ -5,141 +5,148 @@
...
+            std::ofstream output(location, std::ios::binary);
+            std::streampos position;
+
+            writeBinaryPOD(output, maxelements_);
+            writeBinaryPOD(output, size_per_element_);
+            writeBinaryPOD(output, cur_element_count);
+
+            output.write(data_, maxelements_ * size_per_element_);
+
+            output.close();
```
Comment (JoanFM) [inline include/hnswlib/bruteforce.h:124]: u need to cover with `try` and `catch` errors, or u may have an open stream in bad shape

File: include/hnswlib/bruteforce.h [hunk 2]
```diff
@@ -5,141 +5,148 @@
...
+            fstdistfunc_ = s->get_dist_func();
+            dist_func_param_ = s->get_dist_func_param();
+            size_per_element_ = data_size_ + sizeof(labeltype);
+            data_ = (char *) malloc(maxelements_ * size_per_element_);
+            if (data_ == nullptr)
+                std::runtime_error("Not enough memory: loadIndex failed to allocate data");
+
+            input.read(data_, maxelements_ * size_per_element_);
+
+            input.close();
```
Comment (JoanFM) [inline include/hnswlib/bruteforce.h:147]: make sure to handle exceptions to close properly

**Review by JoanFM — CHANGES_REQUESTED** [review]

**Comment by OrangeSodahub** [comment]
In `space_ip.h` and `space_l2.h`, some new functions added. I add the parameters like `const local_state_t *local_state` like other functions to make the parameter lists non-conflicting.

File: include/hnswlib/space_ip.h [hunk 3]
```diff
@@ -3,289 +3,379 @@
...
+            res += ((float *) pVect1)[i] * ((float *) pVect2)[i];
+        }
+        return res;
+
+    }
+
+    static float
+    InnerProductDistance(const void *pVect1, const void *pVect2, const void *qty_ptr, const local_state_t *local_state) {
+        return 1.0f - InnerProduct(pVect1, pVect2, qty_ptr, local_state);
+    }
```
Comment (OrangeSodahub) [inline include/hnswlib/space_ip.h:20]: In `space_ip.h` and `space_l2.h`, some new functions added. I add the parameters like `const local_state_t *local_state` to these new functioins like others to make the parameter lists non-conflicting.

File: include/hnswlib/bruteforce.h [hunk 4]
```diff
@@ -5,141 +5,148 @@
...
+            dist_func_param_ = s->get_dist_func_param();
+            size_per_element_ = data_size_ + sizeof(labeltype);
+            data_ = (char *) malloc(maxElements * size_per_element_);
+            if (data_ == nullptr)
+                std::runtime_error("Not enough memory: BruteforceSearch failed to allocate data");
+            cur_element_count = 0;
+        }
+
+        ~BruteforceSearch() {
+            free(data_);
```
[the remark this answers is not in the document]
Comment (numb3r3) [reply include/hnswlib/bruteforce.h:31]: `bruteforce.h` is never used in annlite now, and will not forever. This file is just a copy from hnswlib.

**Comment by numb3r3** [comment]
@OrangeSodahub FYI, this PR looks good to me. But I would like to hold it until https://github.com/jina-ai/annlite/pull/186 is merged. This PR is a good example to test the wheel building.

**Pull request merged**